File: src/main/java/com/baidu/algorithm/array/_384_Shuffle_an_Array.py
import random


class Solution(object):
    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.nums = nums

    def reset(self):
        """
        Resets the array to its original configuration and return it.
        :rtype: List[int]
        """
        return self.nums

    def shuffle(self):
        """
        Returns a random shuffling of the array.
        :rtype: List[int]
        """
        shuffle = self.nums[:]

        for x in xrange(len(shuffle), 0, -1):
            rand = random.randrange(0, x)
            shuffle[x - 1], shuffle[rand] = shuffle[rand], shuffle[x - 1]

        return shuffle

File: src/main/java/com/baidu/algorithm/datastructure/_384_Shuffle_an_Array.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.datastructure;

import java.util.Random;

/**
 * _384_Shuffle_an_Array
 *
 * @author xuhaoran01
 */
public class _384_Shuffle_an_Array {

    class Solution {

        private int[] origin;
        private int[] array;

        public Solution(int[] nums) {
            origin = nums;
            array = origin.clone();
        }

        /**
         * Resets the array to its original configuration and return it.
         */
        public int[] reset() {
            array = origin;
            origin = array.clone();

            return array;
        }

        /**
         * Returns a random shuffling of the array.
         */
        public int[] shuffle() {
            Random r = new Random();
            for (int i = array.length - 1; i > 0; i--) {
                int j = r.nextInt(i + 1);
                swap(array, i, j);
            }

            return array;
        }

        private void swap(int[] num, int i, int j) {
            int t = num[i];
            num[i] = num[j];
            num[j] = t;
        }
    }
}
